rfcs/text/0000-create-file-mach-o.md

```diff
@@ -0,0 +1,155 @@
+|   file.mach-o.cpu.byte_order               |   keyword  |   CPU byte order for the file.                                              |
+|   file.mach-o.cpu.subtype                  |   keyword  |   CPU subtype for the file.                                                 |
+|   file.mach-o.cpu.type                     |   keyword  |   CPU type for the file.                                                    |
+|   file.mach-o.headers                      |   object   |   Header information for the file.                                          |
+|   file.mach-o.headers.commands             |   object   |   Header load commands for the file.                                        |
+|   file.mach-o.headers.commands.number      |   long     |   Number of load commands for the Mach-O header.                            |
+|   file.mach-o.headers.commands.size        |   long     |   Size of load commands of the Mach-O header.                               |
+|   file.mach-o.headers.commands.type        |   keyword  |   Type of the load commands for the Mach-O header.                          |
+|   file.mach-o.headers.magic                |   keyword  |   Magic field of the Mach-O header.                                         |
+|   file.mach-o.headers.flags                |   keyword  |   Flags set in the Mach-O header.                                           |
+|   file.mach-o.segments                     |   object   |   Segment information for the file.                                         |
+|   file.mach-o.segments.vmaddr              |   keyword  |   Memory address of this segment.                                           |
+|   file.mach-o.segments.name                |   keyword  |   Name of this segment.                                                     |
+|   file.mach-o.segments.vmsize              |   keyword  |   Memory size of this segment.                                              |
+|   file.mach-o.segments.fileoff             |   keyword  |   File offset of this segment.                                              |
+|   file.mach-o.segments.filesize            |   keyword  |   Amount of memory to map from the file.                                    |
+|   file.mach-o.segments.sections            |   object   |   Section information for the segment of the file.                          |
+|   file.mach-o.segments.sections.flags      |   keyword  |   Section flags for the segment of the file.                                |
+|   file.mach-o.segments.sections.name       |   keyword  |   Section name for the segment of the file.                                 |
+|   file.mach-o.segments.sections.type       |   keyword  |   Section type for the segment of the file.                                 |
+|   file.mach-o.signature                    |   object   |   Signature information for the file.                                       |
```

Comment:
I know that there's not necessarily a full corollary with all of the fields that comprise a signed MachO binary, but I'd recommend trying to see if we can use/extend the [`code_signature`](https://www.elastic.co/guide/en/ecs/current/ecs-code_signature.html) fields here--this use case is one of the original reasons for introducing it as a primary field set rather than just having something PE specific.

Reply:
Thanks for the feedback, @andrewstucki 

I looked at [`code_signature`](https://www.elastic.co/guide/en/ecs/current/ecs-code_signature.html) and was able to use `code_signature.status` and remove `file.mach-o.signature.status`. 

Do you think that we should remove `file.mach-o.signature.*` and extend `file.code_signature.*` instead of having signature in two places?

Reply:
Yeah, that would be my suggestion. If possible, try and model some of the sub fields in reusable ways. For example:

I believe`file.mach-o.signature.fingerprint` is related to the x509 cert information embedded in the signature, right? If so, might consider reviving https://github.com/elastic/ecs/pull/852 that I had initially taken a stab at, but then stalled out awhile ago. Additionally for `file.mach-o.signature.team_identifier` is that about the Identifier contained inside the MachO code directory? The description seems to indicate that it's from the code signing cert itself, which I would say could even map to the subject of the cert itself, or a CN in the x509 cert. If it's the code directory identifier, see below around `candidate_cd_hash`.

I don't entirely remember how some of the other pieces of things are encoded in the signature (i.e. `cms_digest`, etc.), but if they're generic ASN.1 attributes, then maybe we could figure out how to model that as a part of generic pkcs7 fields under `code_signature` as well?

For something like `candidate_cd_hash` since this is part of the signature verification process, but not an artifact of a standardized cryptographic signature scheme, I'm wondering if it makes sense just to make a `macho.cdhash` similar to what was suggested for `pe.authentihash`? Similarly for modeling a team identifier pulled from the code directory, something like `macho.signer` or `macho.team_identifier`?

Reply:
**Fingerprint**
Okay, in doing some research ([1](https://objective-see.com/blog/blog_0x26.html), https://github.com/elastic/ecs/pull/852), I would agree that `file.mach-o.signature.fingerprint` is related to the x509 certificate information. I have removed that.

**Team Identifier**
WRT to `file.mach-o.signature.team_identifier` - _Team identifiers are automatically recorded in signatures when signing with suitable Apple-issued signing certificates_ ([2](https://www.manpagez.com/man/1/codesign/osx-10.10.php)). **IMO, I named this improperly**. It appears Team Identifier and Team ID are not the same things (/smh). The data we had shows `11A1A1AAAA` (or the like), which is actually the Team ID vs. the Team Identifier. What about renaming this to `team_id` and align with what it actually is? `team_identifier` does match up nicely to `x509.subject.common_name` and would make sense to be moved there.

Example: 
```
Identifier:  com.cisco.anyconnect.gui (0x60)
Team ID:     DE8Y96K9QP (0x79)
```
**CMS Digest**
`cms_digest` appears to just fall under the [RFC](https://tools.ietf.org/html/rfc5652) and in browsing through the RFC, it appears to be [ASN.1 attributes](https://tools.ietf.org/html/rfc5652#page-43); so where that would need to go in `code_signature` I can't comment on, but I'd agree it's not a Mach-O-specific attribute, then it shouldn't be nested under `file.mach-o.`. I have removed this.

**CD Hash**
I'd agree how you describe it that we should move it up a tier and out of nested underneath `file.mach-o.signature.`. I'll do that.

**Mach-O Signer/Team Identifier**
I noted this above, but `team_identifier` makes sense to be moved to x509.  **Would you recommend moving `team_id` up to `file.mach-o.team_id` from `file.mach-o.signature.team_id`?**

Reply:
Just pinging on this
> Would you recommend moving team_id up to file.mach-o.team_id from file.mach-o.signature.team_id?

Reply:
Sorry, was out for a couple of weeks and then got wrapped up in some other work, so this fell off my radar. In possible, yes.

My thought is that we already have a `code_signature` field set, so rather than introducing a separate `macho.signature` set of fields we could ideally just use `code_signature` nested under `macho` and move any of the `macho`-specific extensions directly into the `macho` namespace. If that doesn't make sense or is a majority of the fields that would exist under `macho.signature` I'm amenable to an argument that `macho.signature` should be a thing. But overall, I think it makes sense to try and re-use existing fields when we can.

Reply:
No worries at all!

Yeah, that makes sense, so I'll remove it here and make a note to look at adding it to [`code_signature`](https://www.elastic.co/guide/en/ecs/current/ecs-code_signature.html).